### PR TITLE
Add Design guidelines

### PR DESCRIPTION
**Pull request by lawli3t**
## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: doc/DESIGN.md [hunk 1]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Memory Management should be handled by the function wherever possible
+- [ ] Functions, who require a large amount of memory to be allocated,
+      state so in their documentation
+
+### Extensibility
+
+- [ ] Function is easily extensible with additional functionality via bitmap-flags
+- [ ] Documentation does not impose limits, that would hinder further extensions
+
+### Tests
```
Comment (lawli3t) [inline doc/DESIGN.md:380]: I am still very undecided on this block. From my understanding it should be possible to judge the bullet points by just looking at the signature and the documentation of a function - no reading of source code whatsoever. For this part one would have to read all the existing tests for a given function, which would require a lot of effort compared to the rest.

Comment (markus2330) [reply doc/DESIGN.md:380]: I am afraid that this step is unavoidable so that we can know that our tests (the real behavior) is consistent to the docu (the expected behavior). Usually a review of the whole implementation is needed to ensure this, we already make a shortcut by only looking at the tests.

File: doc/DESIGN.md [hunk 2]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+- [ ] Added functions are fully covered by tests
+- [ ] All possible error states are covered by tests
+- [ ] All possible configuration options are covered by tests
+- [ ] Tests cover possible edge-cases
```
Comment (markus2330) [inline doc/DESIGN.md:385]: Not needed as separate point, looking at coverage is already enough.

File: doc/DESIGN.md [hunk 3]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] No inconsistencies between tests and code
```
Comment (markus2330) [inline doc/DESIGN.md:386]: As we do not do a full code review, we actually cannot know this.

File: doc/DESIGN.md [hunk 4]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] New API is easily extensible with additional functionality
+- [ ] Components do not depend too heavily on each other
+
+
+## Checklist for each function
+
+### Documentation
+
+- [ ] Change is mentioned in the Compatibility section of the release notes
+- [ ] No inconsistencies between documentation and code
```
Comment (markus2330) [inline doc/DESIGN.md:323]: As we do not do a full code review, we actually cannot know this.

**Review by markus2330 — COMMENTED** [review]
Thank you, great work! I think as next step we should try to go over the checklist together. This way we can probably clarify the points more easily.

File: doc/DESIGN.md [hunk 5]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+      wherever sensible
+- [ ] Wherever possible, function parameters should be `const`
+- [ ] Functions should have the least amount of parameters feasible
+
+### Structural Clarity
+
+- [ ] Function should do exactly one thing
+- [ ] Function should have no side effects
+- [ ] Configuration options are named constants
+- [ ] Function belongs to appropriate group
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply doc/DESIGN.md:367]: Doxygen group

File: doc/DESIGN.md [hunk 6]
```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] ABI/API changes are forward-compatible (breaking backwards-compatibility
+      to add additional symbols is fine)
+
+### Parameter & Return Types
+
+- [ ] Functions should return the most specific type possible
+- [ ] Functions should require the most liberal type possible
+- [ ] Functions should use constant types instead of boolean types
+      wherever sensible
+- [ ] Wherever possible, function parameters should be `const`
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply doc/DESIGN.md:360]: No, I will add it - have you seen any other instances where this might be warranted?

[the remark this answers is not in the document]
Comment (markus2330) [reply doc/DESIGN.md:360]: I think I saw, better you go through all "parameters"

File: scripts/api_review/generate_review_files.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import re
```
Comment (lawli3t) [inline scripts/api_review/generate_review_files.py:1]: wanted to still further improve this - will probably move the method parsing logic out and generate a .json file with it that is then used by this script. This makes it also possible to supply additional info such as markdown-links for docs / definition

File: doc/api_review/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# Elektra API Review
+This folder contains all API design reviews conducted on the Elektra library.
```
Comment (markus2330) [inline doc/api_review/README.md:2]:
```suggestion
This folder contains all API design reviews conducted on LibElektra.
```

File: doc/api_review/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+## Review Process
+This review is performed for each function separately. Each function should be 
+evaluated according to the checklist. The reviewer judges for every bullet 
+point, whether the function does / does not fulfill the bullet point. For 
+every unfulfilled point a short explanation has to be given why the reviewer 
+thinks the function does not fulfill the respective bullet point. This is done 
+simply by providing a short description of the issue below the respective 
+bullet point. After the review has been completed, the reviewer creates an 
+issue for every unfulfilled bullet point so it can be fixed by the Elektra team.
```
Comment (markus2330) [inline doc/api_review/README.md:13]:
```suggestion
issue for every unfulfilled bullet point so it can be fixed by the Elektra Initiative.
```

File: doc/api_review/core/kdbClose.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
+- start = 2021-01-23 18:10
+- end = 2021-01-23 18:10
+- reviewer = Stefan Hanreich <stefanhani@gmail.com>
+
```
Comment (markus2330) [inline doc/api_review/core/kdbClose.md:4]: synopsis and/or version will be needed as the same function name can occur several times due to symbol versioning

File: doc/api_review/core/kdbClose.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+## Checklist
+
+### Documentation
+
+- [ ] Change is mentioned in the Compatibility section of the release notes
+- [ ] No inconsistencies between documentation and tests
+- [ ] Proper Documentation of thread-safety of function
```
Comment (markus2330) [inline doc/api_review/core/kdbClose.md:12]:
```suggestion
- [ ] Proper documentation of thread-safety of function
```

File: doc/api_review/core/kdbClose.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+- [ ] All possible error states are documented
+- [ ] Configuration flags are documented
+
+#### Doxygen
+
+- [ ] Precondition / Postcondition / Invariant
```
Comment (markus2330) [inline doc/api_review/core/kdbClose.md:18]: better put these as several points

File: doc/api_review/core/kdbClose.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+### Naming
+
+- [ ] Abbreviations used in parameter names must be defined in the
+      [Glossary](/doc/help/elektra-glossary.md)
+- [ ] Abbreviations used in function names must be defined in the
+      [Glossary](/doc/help/elektra-glossary.md)
+- [ ] Parameter names should neither be too long, nor too short
+- [ ] Function names should neither be too long, nor too short
+- [ ] Function name should be clear and unambiguous
```
Comment (markus2330) [inline doc/api_review/core/kdbClose.md:33]: Singular/plural (name vs. names, see line above)

File: scripts/api_review/README.md [hunk 14]
```diff
@@ -0,0 +1,2 @@
+# Elektra API Review Template Script
+This python script generates new review files from the given template.
```
Comment (markus2330) [inline scripts/api_review/README.md:2]: You can put this line directly as comment at the top of scripts/api_review/generate_review_files.py

File: scripts/api_review/generate_review_files.py (hunk 7 above)
Comment (markus2330) [reply scripts/api_review/generate_review_files.py:1]: I wouldn't do such things, better keep it as simple as it is.

File: scripts/api_review/template.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+- [ ] Function is easily extensible with function flags (constants)
+- [ ] Documentation does not impose limits, that would hinder further extensions
+
+### Tests
+
+- [ ] Added functions are fully covered by tests
+- [ ] All possible error states are covered by tests
+- [ ] All possible configuration options are covered by tests
+- [ ] Function body has full test coverage
```
Comment (markus2330) [inline scripts/api_review/template.md:75]: Adding some links will be useful during review, e.g. here to the test coverage.

**Review by markus2330 — COMMENTED** [review]
Thank you! Let us look at the details together in a review. I accidentally made suggestions on a generated file.

File: scripts/api_review/template.md [hunk 16]
```diff
@@ -1,36 +1,45 @@
...
 - end = 2021-01-23 18:10
-- reviewer = Stefan Hanreich <stefanhani@gmail.com>
-
+- moderator = Stefan Hanreich <stefanhani@gmail.com>
 
-## Checklist
 
-### Documentation
+## Signature
+{signature}
```
Comment (markus2330) [inline scripts/api_review/template.md:9]: Sometimes there are newlines after the headings, sometimes not. Did you use the `scripts/dev/reformat-markdown` script?

File: scripts/api_review/template.md [hunk 17]
```diff
@@ -1,36 +1,45 @@
...
 
-- [ ] Precondition / Postcondition / Invariant
-- [ ] `@retval`
+- [ ] first line explains briefly what the function does
 - [ ] `@see`
 - [ ] `@since`
 - [ ] `@ingroup`
-- [ ] there is at least one example or snippet how to use the function
+- [ ] `@retval`
+- [ ] good example or snippet how to use the function
```
Comment (markus2330) [inline scripts/api_review/template.md:20]:
```suggestion
- [ ] Good example or snippet how to use the function
```

File: scripts/api_review/template.md [hunk 18]
```diff
@@ -1,36 +1,45 @@
...
+- [ ] simple examples go first
```
Comment (markus2330) [inline scripts/api_review/template.md:21]:
```suggestion
- [ ] Simple examples go first
```

**Review by markus2330 — APPROVED** [review]
Looks very good now!

File: scripts/api_review/template.md (hunk 16 above)
Comment (lawli3t) [reply scripts/api_review/template.md:9]: didn't know this existed - will do!

Comment (lawli3t) [reply scripts/api_review/template.md:10]: ran it now - doesnt seem to have changed anything. not sure if my md is included.
I have adjusted the newlines manually for now though.

**Comment by lawli3t** [comment]
Havent added the reviews for `keyCopy()` and `kdbOpen()` - will add them on the respective branches as soon as they are rebased.

Should now be formatted via reformat-markdown as well, therefore finally ready to go

**Comment by lawli3t** [comment]
Build seems to be still failing due to a reference in Markdown to https://build.libelektra.org/job/elektra-icheck

I've tried to look for the real URL to this job, but can't seem to find it, would you be so nice to point me to the correct one?

**Comment by markus2330** [comment]
There is no separate icheck job anymore, it is integrated in the main (libelektra) pipeline. So https://build.libelektra.org/job/libelektra/job/master/ would be the link.

**Comment by lawli3t** [comment]
> There is no separate icheck job anymore, it is integrated in the main (libelektra) pipeline. So https://build.libelektra.org/job/libelektra/job/master/ would be the link.

Alright, is fixed now via 192e35e4fafddb284414e2073301c4e857578476 - seems to be finally ready now.

I have also added the peer reviews from today

**Comment by markus2330** [comment]
There are still plenty of build jobs failing. Furthermore, please make sure to go though the list in the top-post of the PR and tick appropriately.

**Comment by lawli3t** [comment]
I've added commits that added some additional breakage afterwards, I wasn't aware of the spell checking. All of those issues should be fixed now.

I have looked into the rest and the Jenkins failure is because of missing release notes, those have been added to the respective Markdown-File already - should I just change it to trigger a positive check?

One check fails because [this](https://www.sciencedaily.com/releases/2005/11/051103080801.htm) link is allegedely broken, I cant seem to see why though.

For the other two failures (ASAN, FreeBSD) - i cannot make any sense of the output why CMake would fail. Especially because this PR only consists of Markdown-Changes. Is it possible to download the CMake Logs as artifacts? I havent found a way to do so.

**Comment by markus2330** [comment]
> I have looked into the rest and the Jenkins failure is because of missing release notes, those have been added to the respective Markdown-File already - should I just change it to trigger a positive check?

If you have nothing to add there, you can do that. Without changing the release notes the build server aborts very early and is checking nearly nothing. (This is also why you did not get any feedback about wrongly formatted markdown files.)

> One check fails because this link is allegedely broken, I cant seem to see why though.

You can add the link to ./tests/linkchecker.whitelist

> For the other two failures (ASAN, FreeBSD) - i cannot make any sense of the output why CMake would fail.

Is this maybe what @robaerd is working on in #3654?

**Comment by robaerd** [comment]
> Is this maybe what @robaerd is working on in #3654?

Yes I'm currently working on fixing this issue.

File: doc/DESIGN.md [hunk 19]
```diff
@@ -283,3 +283,104 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+### Compatibility
+
+- [ ] All bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
+- [ ] Components only depend on each other if needed
+
+## Checklist for each function
```
Comment (markus2330) [inline doc/DESIGN.md:316]: Do we need this duplicate? What about simply referring to scripts/api_review/template.c.md

File: scripts/api_review/template.c.md [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+- [ ] Documentation does not impose limits, that would hinder further extensions
+
+### Tests
+
+- [ ] Function code is fully covered by tests
+- [ ] All possible error states are covered by tests
+- [ ] All possible enum values are covered by tests
+- [ ] No inconsistencies between tests and documentation
+
+## Summary
```
Comment (markus2330) [inline scripts/api_review/template.c.md:81]: Maybe also `## Other Issues` or similar here to make counting the lines for your metrics somewhat useful?

**Review by markus2330 — APPROVED** [review]
Looks very good now. Anything left over? If not you can set "ready to merge". If you do any other change, please also do a rebase: then the build should finally pass.

File: doc/DESIGN.md (hunk 19 above)
Comment (lawli3t) [reply doc/DESIGN.md:316]: Added references to the respective templates

File: scripts/api_review/template.c.md (hunk 20 above)
Comment (lawli3t) [reply scripts/api_review/template.c.md:81]: have added it to the templates as well as adjusted the existing reviews

**Comment by lawli3t** [comment]
Have added your suggestions and rebased - build works :tada:

File: scripts/api_review/generate_review_files.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+with open(args.template_path) as template_file:
+    template = template_file.read()
+
+for method in methods:
+    method_name = get_method_name(method)
+
+    with open(get_file_name(method_name), 'w') as output_file:
+        output_file.write(
+            template.format(signature = method, name = method_name)
+        )
```
Comment (mpranj) [inline scripts/api_review/generate_review_files.py:86]: Missing newline at end
```suggestion
        )

```

**Review by mpranj — APPROVED** [review]
Looks good, I committed some minor changes from the review.

**Comment by mpranj** [comment]
Thank you, great work! I look forward to the results of the much-needed API review. 👍
